### PR TITLE
cephadm: Adding poststop actions and setting TimeoutStartSec to 200s

### PR DESCRIPTION
**Pull request by rkachach**
As described in the issue [58242](https://tracker.ceph.com/issues/58242) when a service takes longer than `TimeoutStartSec` cephadm loses the control of the service daemons and there's a mismatch in the information provided by `ceph -s` and `ceph orch ps`. This happens because:

1. `killMode=none` kill mode is set to none in cephadm services (https://github.com/ceph/ceph/pull/33162)
2. cephadm uses the information provided by systemd services status

Basically what happens is that systemd once `TimeoutStartSec` is fired, marks the service as failed (in error) but it doesn't stop the underlying podman container (because `killMode=none`). So from the point of view of cephadm the service daemon is in error but `ceph -s` reports a healthy cluster as in fact daemons ar running (even out of cephadm control).

This PR introduced two changes:

1. Increasse the `TimeoutStartSec` to 200s so in case of really a slow system we give it some more time to start
2. Replicate the same actions from `unit.stop` in `unit.poststop`. This way, even in case of a failing service systemd will stop the underlying podman container (running out of control).

**Note 1:**
This PR keeps the relative order when executing `unit.stop` and `unit.poststop`

**Note 2:**
The reasoning behind adding `unit.poststop` is because these are always executed when a service is restarted by `systemctl restart <service-name>` regardless the current status of the service. `unit.stop` actions are only executed during a restart when the service is running. This can be seen in the following logs (from journalctl):

==> Restarting a service in `failed` status:

```
 Changed failed -> dead
 Changed dead -> start-pre
 Changed start-pre -> start
 Executing: /bin/bash /var/lib/ceph/<fsid>/crash.ceph-node-1/unit.poststop
 Changed start -> stop-post
 Changed stop-post -> failed
```

==> Restarting a service in `running` status:
```
 Changed running -> stop
 Executing: /bin/bash -c 'bash /var/lib/ceph/<fsid>/mgr.ceph-node-0.reytqo/unit.stop'
 Changed stop -> stop-post
 Executing: /bin/bash /var/lib/ceph/<fsid>/mgr.ceph-node-0.reytqo/unit.poststop
 Changed stop-post -> dead
 Changed dead -> start-pre
 Changed start-pre -> start
 Executing: /bin/bash /var/lib/ceph/<fsid>/mgr.ceph-node-0.reytqo/unit.run
 Changed start -> running
```

**PR Testing**
I tested these changes by simulating a slow startup. This could be done by reducing the `TimeoutStartSec` to a very small amount (2sec in my testing). Without the changes on this PR services goes and stays in error state but with the enhancements (specially the ones from the `unit.poststop`) services end up in running state.

Fixes: https://tracker.ceph.com/issues/58241
Fixes: https://tracker.ceph.com/issues/58242

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by adk3798** [comment]
jenkins test make check

**Comment by adk3798** [comment]
For specifically the timeout portion of this PR. Will need to make a new build with the new changes around the post-stop to fully test this

-----

https://pulpito.ceph.com/adking-2023-01-21_05:38:21-orch:cephadm-wip-adk-testing-2023-01-20-1359-distro-default-smithi/

Lots of failures (13) but all accounted for

- 3 instances of `'list' object has no attribute 'get'`. Failure was on new test for setting mon crush location introduced by a PR in the run. That PR can't be merged, but this shouldn't block merging anything else.
- 2 instances of test_cephadm.sh test failing, caused by the same mon crush location PR. Failed with `local variable 'config_json' referenced before assignment`
- 1 instance of https://tracker.ceph.com/issues/49287, known issue
- 2 instances of https://tracker.ceph.com/issues/58524, known issue
- 2 instances of https://tracker.ceph.com/issues/58535, known issue
- 1 instance of https://tracker.ceph.com/issues/57771, known issue
- 1 instance of https://tracker.ceph.com/issues/57755, known issue
- 1 failed basic smoke test that timed out waiting for all 8 OSDs to come up. Did 5 reruns of this test with the same build (https://pulpito.ceph.com/adking-2023-01-22_15:28:20-orch:cephadm-wip-adk-testing-2023-01-20-1359-distro-default-smithi/) all of which passed. My guess is an I/O error creating the last OSD, which I've seen happen (infrequently) when we make a bunch of OSDs on the split up nvme drive we use for having multiple OSDs on these machines. Considering it a 1 off for now, and not blocking merging over it.

Overall, PRs in the run should be okay to merge other than the mon crush location one causing failures. Will start to try and clean up the test suite now that we're able to make builds and run tests again.

**Comment by adk3798** [comment]
This time tested with the post-stop changes included

----

https://pulpito.ceph.com/adking-2023-01-22_23:00:58-orch:cephadm-wip-adk-testing-2023-01-22-1523-distro-default-smithi/

lots of failures from infra stuff (re-imaging machines, installing things pre-test) so did a rerun of all failed and dead jobs, resulting in

https://pulpito.ceph.com/adking-2023-01-23_04:36:45-orch:cephadm-wip-adk-testing-2023-01-22-1523-distro-default-smithi/

Leaving us with 6 failures.

- 3 failed jobs were the new `test_set_mon_crush_locations` test introduced in one of the PRs in the run. The test just isn't quite right yet, but it's not relevant in terms of whether we can merge other PRs
- 2 instances of the staggered upgrade test issue (known issue) https://tracker.ceph.com/issues/58535
- 1 instance of failing to install container selinux package (known issue) https://tracker.ceph.com/issues/57771

Overall, nothing that would block merging for anything other than the mon crush location PR.

File: src/cephadm/cephadm.py [hunk 1]
```diff
@@ -3511,10 +3520,7 @@ def deploy_daemon_units(
         # following generated script basically checks if the container exists
         # before stopping it. Exit code will be success either if it doesn't
         # exist or if it exists and is stopped successfully.
```
Comment (mgfritch) [inline src/cephadm/cephadm.py:3522]: (nit) move this comment into the new `add_stop_action` func?

**Review by mgfritch — APPROVED** [review]

File: src/cephadm/cephadm.py (hunk 1 above)
Comment (rkachach) [reply src/cephadm/cephadm.py]: @mgfritch you are right, moved :+1:

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-02-13_21:57:03-orch:cephadm-wip-adk-testing-2023-02-13-1359-distro-default-smithi/

11 failures, one dead job

- 8 of the failed jobs were on new tests being added in by other PRs in the run, which obviously aren't quite working yet (test_monitoring_stack_basic, rotate-ssh-keys, test_iscsi_container) but outside of iscsi and monitoring stack specific changes and the PRs that introduced the tests, this shouldn't block merging.
- 2 instances of https://tracker.ceph.com/issues/58535
- 1 instance of https://tracker.ceph.com/issues/49287
- 1 dead job was an error re-imaging machines, infra issue

**Pull request merged**